src/util.rs

```diff
@@ -4,15 +4,20 @@ use std::{io, iter::repeat_with};
 
 use crate::error::IoResultExt;
 
-fn tmpname(prefix: &OsStr, suffix: &OsStr, rand_len: usize) -> OsString {
-    let mut buf = OsString::with_capacity(prefix.len() + suffix.len() + rand_len);
+fn tmpname(prefix: &OsStr, suffix: &OsStr, rand_len: usize) -> io::Result<OsString> {
+    let capacity = prefix
+        .len()
+        .checked_add(suffix.len())
+        .and_then(|value| value.checked_add(rand_len))
+        .ok_or(io::Error::new(io::ErrorKind::InvalidInput, "name too long"))?;
```

Comment:
Technically, the rand_len just has to be large enough, which just requires a `.rand_bytes(usize::MAX)` call. No need to a huge allocation of a filename beforehand.

But yes, you could argue that it's a user error to call that function with such a large value.

Reply:
(well, it'll likely run out of memory and panic if the length is anywhere near usize::MAX)

But... how about we just use saturating math here as well (this is just for capacity). It avoids the error case and, and avoids the `and_then` etc.

Reply:
Yeah, why not? I wanted to avoid the panic, but if someone really wants to try to allocate THAT much or just more than available, standard Rust will run into a panic anyway. The user asked for trouble doing that. :)

Adjusted as well. And removed the test.